### PR TITLE
Uploading to user's external S3 bucket(SYNPY-179)

### PR DESCRIPTION
**Pull request by zimingd**

### CODE REVIEW

File: synapseclient/client.py [hunk 1]
```diff
@@ -126,6 +126,12 @@
                          "max_wait"          : 30,
                          "back_off"          : 2}
 
+#Concrete types for UploadDestinations
+_SYNAPSE_S3_UPLOAD_DESTINATION_TYPE = 'org.sagebionetworks.repo.model.file.S3UploadDestination'
+_EXTERNAL_UPLOAD_DESTINATION_TYPE = 'org.sagebionetworks.repo.model.file.ExternalUploadDestination'
+_EXTERNAL_S3_UPLOAD_DESTINATION_TYPE = 'org.sagebionetworks.repo.model.file.ExternalS3UploadDestination'
+
```
Comment (larssono) [inline synapseclient/client.py]: Why separate these as constants?  We have been using the concrete type all over the code before - is the plan that we define all of these as constants eventually?

File: synapseclient/client.py [hunk 2]
```diff
@@ -1988,78 +1996,62 @@ def _deleteFileHandle(self, fileHandle):
     ############################################################
 
     def _getDefaultUploadDestination(self, entity):
-        return self.restGET('/entity/%s/uploadDestination'% id_of(entity),
+        return self.restGET('/entity/%s/uploadDestination'% entity['parentId'],
                      endpoint=self.fileHandleEndpoint)
```
Comment (larssono) [inline synapseclient/client.py]: Why the change to parentId?  Was this a bug?  I.e.  did this function not actually work?

File: synapseclient/client.py [hunk 3]
```diff
@@ -1988,78 +1996,62 @@ def _deleteFileHandle(self, fileHandle):
...
         It not an external storage location returns the original path.
 
         :param entity: An entity with path.
 
-        :returns: A URL or local file path to add to Synapse along with an update local_state
-                  containing externalURL and content-type
+        :returns: A URL or local file path to add to Synapse,
+                  an update local_state containing externalURL and content-type,
+                  and a storage location id indicating to where the entity should be uploaded ("None" defaults to Synapse)
         """
```
Comment (larssono) [inline synapseclient/client.py]: This is insconsistent with the function header.  The local_state is not returned but is a parameter.

Since we are changing this function we can advantage of the refactor to rename it _uploadExternalllyStoringProjects (_ _ functions are generally a bad idea).

File: synapseclient/client.py (hunk 1 above)
Comment (zimingd) [inline synapseclient/client.py]: Move these concrete types into a constants file: concreteTypes.py

File: synapseclient/client.py [hunk 4]
```diff
@@ -1988,78 +1996,62 @@ def _deleteFileHandle(self, fileHandle):
...
+                if local_state.get('contentType') is None:
+                    mimetype, enc = mimetypes.guess_type(entity['path'], strict=False)
+                    local_state['contentType'] = mimetype
+                return uploadLocation, local_state, location['storageLocationId']
+            else:
+                raise NotImplementedError('Can only handle SFTP upload locations.')
         else:
-            raise NotImplementedError('Can only handle S3 and SFTP upload locations.')
-
+            raise NotImplementedError("The UploadDestination type: " + upload_destination_type + " is not supported")
```
Comment (zimingd) [inline synapseclient/client.py]: strip the org.sagebionetworks.....

File: synapseclient/client.py [hunk 5]
```diff
@@ -2864,6 +2856,8 @@ def _uploadCsv(self, filepath, schema, updateEtag=None, quoteCharacter='"', esca
         :param filepath: Path of a `CSV <https://en.wikipedia.org/wiki/Comma-separated_values>`_ file.
         :param schema: A table entity or its Synapse ID.
         :param updateEtag: Any RowSet returned from Synapse will contain the current etag of the change set. To update any rows from a RowSet the etag must be provided with the POST.
+        :param storageLocationId: a id indicating where the file should be stored. retrieved from Synapse's UploadDestination
```
Comment (zimingd) [inline synapseclient/client.py]: remove

File: tests/integration/integration_test.py [hunk 6]
```diff
@@ -389,3 +389,70 @@ def test_teams():
 
     assert team == found_teams[0]
 
+def _set_up_external_s3_project():
```
Comment (zimingd) [inline tests/integration/integration_test.py]: skip test if not synapse-test

File: tests/integration/integration_test.py [hunk 7]
```diff
@@ -389,3 +389,70 @@ def test_teams():
...
+    project_destination = {'concreteType': 'org.sagebionetworks.repo.model.project.UploadDestinationListSetting',
+                           'settingsType': 'upload',
+                           'locations': [destination['storageLocationId']],
+                           'projectId' : project_ext_s3.id}
+
+    project_destination = syn.restPOST('/projectSettings', body=json.dumps(project_destination))
+    schedule_for_cleanup(project_ext_s3)
+    return project_ext_s3.id, destination['storageLocationId']
+
+def test_external_s3_upload():
```
Comment (zimingd) [inline tests/integration/integration_test.py]: skip this test if not logged in as synapse-test

File: tests/integration/integration_test.py [hunk 8]
```diff
@@ -389,3 +389,70 @@ def test_teams():
...
+    #setup
+    project_id, storage_location_id = _set_up_external_s3_project()
+
+    # create a temporary file for upload
```
Comment (zimingd) [inline tests/integration/integration_test.py]: use convenience function instead: make_bogus_data_file()
